### PR TITLE
Decreasing the overallocation for list aggregates

### PR DESCRIPTION
**Pull request by taniabogatsch**
This PR addresses the problem described in #3734. This query was causing DuckDB to run out of memory.
```sql
SELECT tconst, list(principals) as principals FROM test.parquet GROUP BY 1 LIMIT 10
```

The table has this schema. There are `8074739` distinct groups.
```
tconst: VARCHAR
principals: STRUCT(
   ordering INTEGER, 
   nconst VARCHAR, 
   category VARCHAR, 
   job VARCHAR, 
   characters []VARCHAR)
```
If it is run with e.g. 8 threads, then each thread builds a hash table with ~8mio. `LIST` vectors. Each of these vectors contains a `STRUCT` of `VARCHARs` and even another `LIST` of `VARCHARs`. As described by @lnkuiper this caused a massive overallocation, because each vector was allocated with a size of `STANDARD_VECTOR_SIZE = 1024`. DuckDB was not actually leaking memory, it was running out of it.

This PR allocates each of these vectors with an initial capacity of 1. This allows the query to succeed on my laptop.

Even though this should fix the issue, there are still several issues, as already mentioned in #3734 by Laurens. 
- The resulting list of structs contains strings that exceed the length of 12. DuckDB creates a `StringHeap` to store these, which is again done in each thread for each list vector. `StringHeap` has a fixed size of `MINIMUM_HEAP_SIZE = 4096`. Reducing this size to 512 already reduces the memory consumption by half. So while this PR tackles the problem of over allocating vectors, the problem of over allocating the string heaps persists.
- When profiling the query, it can be seen that while the `ListUpdateFunction` runs in parallel, `ListCombineFunction`, `ListFinalize` and the cleanup are performed by only one thread
   - For `ListCombineFunction` that means that one thread has to combine the results of e.g. 8 different threads, each containing approx. 8mio. list vectors
   - Since the cleanup is also performed by only one thread, it takes much longer than the actual query. A possible improvement would be to already start cleaning up the results of one thread once they got combined to the final result.

To sum up, there is a lot of room for improvements concerning the `LIST` aggregate function that will take some effort to implement. @Mytherin and @lnkuiper I am not sure how important this is, though, because this query is (so far) quite special.

### CODE REVIEW

**Review by lnkuiper — CHANGES_REQUESTED** [review]
Just a minor nitpick, then it looks good to me! We talked about what to do with the `StringHeap`, and came to the conclusion that we probably need one `StringHeap` per thread in the aggregation, which should store strings in buffer-managed blocks. This will greatly reduce the number of (de)allocations, which will deal with some of the problems mentioned.

**Review by hawkfish — COMMENTED** [review]
Can we add a test that forces the lists to reallocate?

**Comment by Mytherin** [comment]
Thanks!
